src/asgen/backends/alpinelinux/apkindexutils.d

```diff
@@ -0,0 +1,149 @@
+        return format ("%s-%s.apk", this.pkgname, this.pkgversion);
+    }
+}
+
+/**
+* Range for looping over the contents of an APKINDEX, block by block.
+*/
+struct ApkIndexBlockRange {
+    this (string contents)
+    {
+        this.lines = contents.splitLines;
+        this.getNextBlock();
+    }
+
+    @property ApkIndexBlock front () const {
+        return this.currentBlock;
+    }
+
+    @property bool empty () {
+        return this.m_empty;
+    }
+
+    void popFront ()
+    {
+        this.getNextBlock ();
+    }
+
+    @property ApkIndexBlockRange save () { return this; }
+
+private:
+    void getNextBlock () {
+        string[] completePair;
+        uint iterations = 0;
+        foreach (currentLine; this.lines[this.lineDelta .. $]) {
+            iterations++;
+            if (currentLine == "") {
+                // next block for next package started
+                break;
+            } if (currentLine.canFind (":")) {
+                if (completePair.empty) {
+                    completePair = [currentLine];
+                    continue;
+                }
+
+                auto pair = completePair.join (" ").split (":");
+                this.setCurrentBlock (pair[0], pair[1]);
+                completePair = [currentLine];
+            } else {
+                completePair ~= currentLine.strip ();
+            }
+        }
+
+        this.lineDelta += iterations;
+        this.m_empty = this.lineDelta == this.lines.length;
+    }
+
+    void setCurrentBlock (string key, string value) {
+        switch (key) {
+            case "A":
+                this.currentBlock.arch = value;
+                break;
+            case "m":
+                this.currentBlock.maintainer = value;
+                break;
+            case "P":
+                this.currentBlock.pkgname = value;
+                break;
+            case "T":
+                this.currentBlock.pkgdesc = value;
+                break;
+            case "V":
+                this.currentBlock.pkgversion = value;
+                break;
+            default:
+                // We dont care about other keys
+                break;
+        }
+    }
+
+    string[] lines;
+    ApkIndexBlock currentBlock;
+    bool m_empty;
+    uint lineDelta;
+}
+
+static assert (isForwardRange!ApkIndexBlockRange);
+
+/**
+ * Download apkindex if required. Returns the path to the local copy of the APKINDEX.
+ */
+immutable (string) downloadIfNecessary (const string prefix,
+                                        const string destPrefix,
+                                        const string srcFileName,
+                                        const string destFileName)
+{
+    auto downloader = Downloader.get;
+
+    immutable filePath = buildPath (prefix, srcFileName);
+    immutable destFilePath = buildPath (destPrefix, destFileName);
+
+    if (filePath.isRemote) {
+        try {
+            /* This should use download(), but that doesn't throw errors */
```

Comment:
That comment is no longer true with the refactored downloader and could just be deleted.